### ISSUE TITLE
gmock issue building realtime_tools on arch linux

### ISSUE THREAD

**Issue by jwhendy** [ros/catkin#1030]
I'm trying to build [ros-melodic-realtime-tools](https://aur.archlinux.org/packages/ros-melodic-realtime-tools/) from the arch linux AUR. [The PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ros-melodic-realtime-tools) shows how the make commands are called.

I get the following error:
```
$ makepkg
==> Making package: ros-melodic-realtime-tools 1.15.0-0 (Mon 09 Sep 2019 01:57:27 AM CDT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found ros-melodic-realtime-tools-1.15.0.tar.gz
==> Validating source files with sha256sums...
    ros-melodic-realtime-tools-1.15.0.tar.gz ... Passed
==> Extracting sources...
  -> Extracting ros-melodic-realtime-tools-1.15.0.tar.gz with bsdtar
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- Using CATKIN_DEVEL_PREFIX: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/melodic
-- This workspace overlays: /opt/ros/melodic
-- Using PYTHON_EXECUTABLE: /usr/bin/python3
-- Using default Python package layout
-- Using empy: /usr/lib/python3.7/site-packages/em.py
-- Using CATKIN_ENABLE_TESTING: ON
-- Skip enable_testing() when building binary package
-- Using CATKIN_TEST_RESULTS_DIR: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-3.7
-- catkin 0.7.17
-- BUILD_SHARED_LIBS is on
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)


CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.


-- Configuring incomplete, errors occurred!
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeOutput.log".
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeError.log".
```

I made a slight edit to a function to see where it was looking.
```
$ cat /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake
function(_catkin_add_executable_with_google_test type target)
...
  if(NOT ${type_upper}_FOUND AND NOT ${type_upper}_FROM_SOURCE_FOUND)
    message(WARNING "skipping ${type} '${target}' in project '${PROJECT_NAME}' because ${type} was not found in '${GTEST_INCLUDE_DIRS}'")
```

This prints out as:
```
skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found in '/usr/include'
```

I do have gmock/gtest installed, and the dirs are in `/usr/include`. For the heck of it, I tried symlinking all `*.h` files into `/usr/include` directly, but that didn't work out.

```
$ pacman -Qi gmock
Name            : gmock
Version         : 1.8.1-3

$ pacman -Qi gtest
Name            : gtest
Version         : 1.8.1-3

$ ls /usr/include/gmock/
internal         gmock-cardinalities.h      gmock-generated-function-mockers.h  gmock-generated-nice-strict.h  gmock-matchers.h      gmock-more-matchers.h
gmock-actions.h  gmock-generated-actions.h  gmock-generated-matchers.h          gmock.h                        gmock-more-actions.h  gmock-spec-builders.h
```

I think this is related to #933 and #970. I also saw #1022  and adapted ros-melodic-catkin to build from @sloretz 's fork, but this did not solve my issue. I get the same original error about gmock not being found.

Any other details that would be helpful, or debug output I could add to assist with this?

**Comment by GladOSkar** [comment]
Same issue on debian 10 buster

This is when building `melodic-desktop-full` from source

Manually installing `libgmock-dev` (which contains the files in `/usr/include/gmock`) did **not** resolve the issue.

I will try applying #1022 later

**Comment by GladOSkar** [comment]
Applying #1022 solved the issue for me

**Comment by jwhendy** [comment]
@GladOSkar which package from `melodic-desktop-full` fails for you? I will retry #1022 but as noted above, when I installed @sloretz 's fork of `catkin` with that change, it didn't work for me. I'll try again in case I goofed something, or perhaps the package failing for you and me (realtime_tools) is not the same?

---
*EDIT*: Another thought is that the #1022 patch fixes the issue on debian, but not arch? If the magic is adding `/usr/include/gmock`, that exists on arch as well. Perhaps post everything from `ls /usr/include/gmock` so we can compare (mine is posted above).

**Comment by GladOSkar** [comment]
@jwhendy it was `realtime-tools` for me as well.

Yeah I saw that #1022 didn't fix it for you, weird. An unusual thing I did is that I applied each patch line manually (didn't use his file or the patch command) because I feared conflict since their patch is for kinetic.

From how I understood their PR I don't think the issue is with /usr/include/gmock, but rather catkin's gmock detection in the cmake file is faulty for newer versions of gmock. Not entirely sure though.

**Comment by jwhendy** [comment]
Any pointers on how best to do this? This repo stops at `kinetic-devel`, so I don't think it's a `melodic` vs. `kinetic` thing.

If I just clone @sloretz 's fork using the `gmock-from-source-buster` branch, should that work? I grepped for the various `googltest` instances in the updated `gtest.cmake` and see them. It appears this is the only thing that's changed. I still fail to find `gmock`.

```
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)
```

**Comment by jwhendy** [comment]
Also, do you actually have these directories, or are they really variables that get resolved?
```
$ grep googletest /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake 
  list(APPEND _gtest_include_paths "${include_path}/googletest/googletest/include/gtest")
  list(APPEND _gtest_source_paths "${src_path}/googletest/googletest/src")
    list(INSERT _gtest_include_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googletest/include/gtest")
    list(INSERT _gtest_source_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googletest/src")
  list(APPEND _gmock_include_paths "${include_path}/googletest/googlemock/include/gmock")
  list(APPEND _gmock_source_paths "${src_path}/googletest/googlemock/src")
    list(INSERT _gmock_include_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googlemock/include/gmock")
    list(INSERT _gmock_source_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googlemock/src")
```

I don't have any `googletest` or `googlemock` directories.

**Comment by jwhendy** [comment]
From looking at those directories, here is arch's layout:
```
$ ls /usr/include/gtest/
internal            gtest-message.h     gtest-printers.h  gtest-test-part.h
gtest-death-test.h  gtest-param-test.h  gtest_prod.h      gtest-typed-test.h
gtest.h             gtest_pred_impl.h   gtest-spi.h

$ ls /usr/include/gmock/
internal                   gmock-generated-function-mockers.h  gmock-matchers.h
gmock-actions.h            gmock-generated-matchers.h          gmock-more-actions.h
gmock-cardinalities.h      gmock-generated-nice-strict.h       gmock-more-matchers.h
gmock-generated-actions.h  gmock.h

$ ls /usr/src/gtest/
cmake  src  CMakeLists.txt

$ ls /usr/src/gmock/
gmock-all.cc            gmock.cc                 gmock_main.cc      gmock-spec-builders.cc
gmock-cardinalities.cc  gmock-internal-utils.cc  gmock-matchers.cc
```

**Comment by GladOSkar** [comment]
> Any pointers on how best to do this?

No idea TBH, i don't know a lot about these things either ¯\_(ツ)_/¯

My `/usr/src` dirs do contain more complete source files than yours:

![image](https://user-images.githubusercontent.com/9497832/64978090-00772800-d8b5-11e9-8404-6dc44d8f97dd.png)

Not sure if this helps

**Comment by jwhendy** [comment]
Hmmm, that's super interesting.
```
$ pacman -Qi gmock
Name            : gmock
Version         : 1.8.1-3

$ pacman -Qi gtest
Name            : gtest
Version         : 1.8.1-3
```

Can you post the same, as well as something like (just a few, you don't need all of these)
```
$ pacman -Qo /usr/src/gmock/*
$ pacman -Qo /usr/src/gtest/*
```
I want to understand why those files are missing for me, but not you.

**Comment by jwhendy** [comment]
Oh, whoops. Bah. I'm collaborating with an Oskar on arch packages and thought you were him. Okay, I think this is a Debian vs. Arch thing, then. I definitely don't have a full src folder like that for either of these packages.

**Comment by GladOSkar** [comment]
Haha no worries :)

Good luck finding a solution!

**Comment by jwhendy** [comment]
Thanks for the post of the directories. I think that's a great start and enabled me to create an arch [bug report](https://bugs.archlinux.org/task/63793). I'll post back when we resolve. I'm not sure where to find a de facto directory structure of what should/should not exist and where...

**Comment by dirk-thomas** [comment]
@jwhendy Any update on this? Btw the referenced PRs have been merged and released in the meantime so you might want to give it another try on Arch.

**Comment by jwhendy** [comment]
@dirk-thomas I haven't done anything since I didn't hear any news here or on the Arch bug.

>Btw the referenced PRs have been merged and released in the meantime so you might want to give it another try on Arch.

Are there others besides #1022? In my first post, I note that I built directly from sloretz's fork to try #1022 directly. I can certainly try again later today.

From my last comment:

>  I'm not sure where to find a de facto directory structure of what should/should not exist and where...

This would be super helpful. I couldn't figure out from the code what, specifically, the cmake check is not finding. Any coaching here? The posts of directory structure from @GladOSkar were potentially promising, as they show Arch doesn't have the same `/lib` and `/src` contents.

If I can nail down what `catkin` expects, I can also feed this to the Arch packagers to make sure things are installing properly.

---

As a final aside, I remembered that I [did this](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L401) without success.

```
if(CATKIN_TOPLEVEL)
    message(STATUS "gtest not found, C++ tests can not be built. Please install the gtest headers globally in your system or checkout gtest (by running 'git clone  https://github.com/google/googletest.git -b release-1.8.0' in the source space '${CMAKE_SOURCE_DIR}' of your workspace) to enable gtests")
  else()
```

Is there a reason this is only proposed for `gtest` vs. also being a workaround if `gmock` is not found?

**Comment by dirk-thomas** [comment]
> Are there others besides #1022?#1021

**Comment by dirk-thomas** [comment]
> Are there others besides #1022?

I was think of #1021 which was also an issue on Buster. Though it might not be relevant here.

> This would be super helpful. I couldn't figure out from the code what, specifically, the cmake check is not finding. Any coaching here?

You can either look at the directory structure on Ubuntu or use the paths which are being checked in the CMake code.

> Is there a reason this is only proposed for `gtest` vs. also being a workaround if `gmock` is not found?

This corner case has never been necessary. It is extremely unlikely now that gtest and gmock are bundled together that you only have gtest but not gmock and want to provide gmock from source within the workspace.

**Comment by jwhendy** [comment]
> I was think of #1021

I'll do a clean build from master tonight. Thanks!

> You can either look at the directory structure on Ubuntu or use the paths which are being checked in the CMake code.

Indeed, above I tried to replicate the [Ubuntu filesystem screenshot](https://github.com/ros/catkin/issues/1030#issuecomment-531868346) posted by @GladOSkar without success.

I've also looked at [the code ](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L180) and it looks like the find script is looking for `gmock.h` and `gmock.cc`. Note that these exist per [my post above](https://github.com/ros/catkin/issues/1030#issuecomment-531825849).

Is anyone able to explain what's going wrong in plain english? So far, we haven't pinpointed what's actually going wrong. Based on the information I've provided, what looks awry? Or, what information would you need to suggest what's going wrong?

**Comment by jwhendy** [comment]
I just re-built and installed `ros-melodic-catkin` looking at `kinetic-devel`, so I should be up to date:

```
### from pkg build directory
$ cat src/ros-melodic-catkin/CHANGELOG.rst | head

0.7.19 (2019-10-08)
-------------------
* support for pytest 5.1+ jUnit result files (`#1033 <https://github.com/ros/catkin/issues/1033>`_)
* fix -pthread handling in Debian buster (`#1021 <https://github.com/ros/catkin/issues/1021>`_)
* gmock from source on Debian buster (`#1022 <https://github.com/ros/catkin/issues/1022>`_)
* [windows] update a typo in local_setup.bat.in (`#1029 <https://github.com/ros/catkin/issues/1029>`_)
```

I get the same build error with `realtime_tools`:

```
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)


CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.
```

And here is my attempt to understand why this fails.

[Here's where](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L268) it looks:
```
_catkin_find_google_source("gmock" "${_gmock_include_paths}"
                             "${_gmock_source_paths}" _gmock_found
                             _gmock_base_dir _gmock_include_dir _gmock_lib_dir
                             _gmock_libs _gmock_main_libs)
```
That [calls this](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L181):
```
function(_catkin_find_google_source type include_paths
  src_paths found base_dir include_dir lib_dir libs main_libs)
  # Find the gtest headers
  find_file(_${type}_INCLUDES "${type}.h"
            PATHS ${include_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)

  # Find the gtest sources
  find_file(_${type}_SOURCES "${type}.cc"
            PATHS ${src_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```

Thus, we look for two files from what I can tell:
```
find_file(_gmock__INCLUDES "gmock.h"
            PATHS ${include_paths}

find_file(_gmock__SOURCES "gmock.cc"
            PATHS ${src_paths}
```

At face value, do we have those files?
```
$ locate gmock.cc
/usr/src/gmock/gmock.cc

$ locate gmock.h
/usr/include/gmock/gmock.h
```

Ok, edit `gtest.cmake` to help us find out what it's looking for and where:
```
# Internal function for finding gtest or gmock sources
function(_catkin_find_google_source type include_paths
  src_paths found base_dir include_dir lib_dir libs main_libs)
  # Find the gtest headers
  message("##### _${type}_INCLUDES ${type}.h ${include_paths}")
  find_file(_${type}_INCLUDES "${type}.h"
            PATHS ${include_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)

  # Find the gtest sources
  message("##### _${type}_SOURCE ${type}.cc ${src_paths}")
  find_file(_${type}_SOURCES "${type}.cc"
            PATHS ${src_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```

What do we get?
```
##### _gtest_INCLUDES gtest.h /usr/include;/usr/src/gtest;/usr/include;/usr/src/googletest/googletest/include/gtest
##### _gtest_SOURCE gtest.cc /usr/src/gtest/src;/usr/src/googletest/googletest/src
##### _gmock_INCLUDES gmock.h /usr/include;/usr/src/gmock;/usr/include;/usr/src/googletest/googlemock/include/gmock
##### _gmock_SOURCE gmock.cc /usr/src/gmock/src;/usr/src/googletest/googlemock/src
```

So, it looks like it's looking or `/usr/include/{gtest,gmock}.h` and `/usr/src/{gtest, gmock}/src/{gtest, gmock}.cc` On arch linux, three were not in the right place according to these paths, but I fixed that with symlinks.
```
$ ls -l /usr/include/gtest.h 
lrwxrwxrwx 1 root root 26 Oct  8 21:30 /usr/include/gtest.h -> /usr/include/gtest/gtest.h
$ ls -l /usr/include/gmock.h
lrwxrwxrwx 1 root root 26 Oct  8 21:28 /usr/include/gmock.h -> /usr/include/gmock/gmock.h

$ ls -l /usr/src/gmock/src/gmock.cc 
lrwxrwxrwx 1 root root 23 Oct  8 21:34 /usr/src/gmock/src/gmock.cc -> /usr/src/gmock/gmock.cc

### gtest gets installed here by default, no need to symlink
$ ls -l /usr/src/gtest/src/gtest.cc 
-rw-r--r-- 1 root root 217839 Oct  1  2018 /usr/src/gtest/src/gtest.cc
```

Ah! Interesting. No error. I didn't expect that, and was pretty sure I did this up above. That said, printing out the includes from that function did leave me puzzled. Note [lines 258-259](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L258):
```
set(_gmock_include_paths "${include_path}/gmock")
set(_gmock_source_paths "${src_path}/gmock/src")
# ...
_catkin_find_google_source("gmock" "${_gmock_include_paths}"
                             "${_gmock_source_paths}" _gmock_found
                             _gmock_base_dir _gmock_include_dir _gmock_lib_dir
                             _gmock_libs _gmock_main_libs)
```

In my `message()` call above from within `function(_catkin_find_google_source type include_paths...)`, it looks to have been passed `include_paths=${_gmock_include_paths}`, which looks to include `${include_path}/gmock`. But only `/usr/include` was printed out. So, this issue looks like two issues:

- arch linux needs to add `/usr/src/gmock/src/{gmock-all.cc, gmock.cc, gmock_main.cc}` to their package.
- I _think_ `/usr/include/{gmock, gtest}/{gmock, gtest}.h` seems like a reasonable location for these files. What about a PR to add those to the include paths in `catkin`?
- I also think it's odd to have `/usr/src/gmock` in `_gmock_include_paths` (if it was in `_gmock_source_paths`, arch's default gmock location would work), and `/usr/include` is in there twice (should one have been `/usr/include/gmock`?). How about a bonus PR to clean that up, if possible?

**Comment by dirk-thomas** [comment]
> I *think* `/usr/include/{gmock, gtest}/{gmock, gtest}.h` seems like a reasonable location for these files. What about a PR to add those to the include paths in `catkin`?

These are the exact paths used by the Debian package `googletest`: see https://packages.ubuntu.com/bionic/amd64/googletest/filelist As far as I can see the headers are already being found in those locations so I don't understand what your PR would contain?

> I also think it's odd to have `/usr/src/gmock` in `_gmock_include_paths`

Maybe that was necessary for versions of googletest which ship only the sources? If you want to remove something like this it would need to be thoroughly tested that it doesn't break any use case (on Ubuntu, on Debian, different version of googletest, from-source, etc.)

> and `/usr/include` is in there twice (should one have been `/usr/include/gmock`?)

Please use references to exact lines of code otherwise I am unable to follow your argument.

> How about a bonus PR to clean that up, if possible?

Please feel free to propose any kind of PR. Only when seeing the actual proposed diff I can give you concrete feedback.

[As mentioned above please make sure to do thorough testing of any changes - often it is hard to justify the necessary testing effort for "just cleaning up something".]

**Comment by jwhendy** [comment]
>As far as I can see the headers are already being found in those locations so I don't understand what your PR would contain?

Well, _something_ is awry, as even after fixing the `*.cc` files, I indeed had the `.h` files in the respective `/usr/include/{gmock, gtest}/{gmock, gtest}.h` locations and they were not found. I can't explain the results on debian.

>Maybe that was necessary for versions of googletest which ship only the sources? 

Maybe, but I think it could be an accident based on findings below.

>Please use references to exact lines of code otherwise I am unable to follow your argument.

I thought I did? The bullet you're commenting on was a closing summary based on the code and output I'd already laid out, including both linked lines and copy/pasted chunks. To summarize again: my observation was that printing out the `include_paths` argument passed to `catkin_find_google_source()` ([here](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L181)) didn't show `/usr/include/gmock`, but that is what `_gmock_include_paths` appears to be set to ([here](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L258)) when it's passed ([here](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L268)).

I did make a potential breakthrough while writing this. Here was an experiment on [this function](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L268) just to make sure the passed path was really what I thought:
```
#Find GMock
  set(_gmock_include_paths "${include_path}/gmock/whammy") ### I changed this line
  set(_gmock_source_paths "${src_path}/gmock/src")
  # ...
  _catkin_find_google_source("gmock" ${_gmock_include_paths}"
                             "${_gmock_source_paths}" _gmock_found
                             _gmock_base_dir _gmock_include_dir _gmock_lib_dir
                             _gmock_libs _gmock_main_libs)
```

I then added output to the [function called](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L181):

```
# Internal function for finding gtest or gmock sources
function(_catkin_find_google_source type include_paths
  # ...
  message("### ${include_paths}")
  message("### ${src_paths}")
```

Output:
```
### /usr/include;/usr/src/gmock/whammy
### /usr/src/gmock/src
```

How is `/usr/src/gmock/whammy` getting in there vs. the expected `/usr/include/gmock/whammy`?

I tracked it down [to this](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L303):
```
if(NOT GMOCK_FOUND OR NOT GTEST_FOUND)
  # ...
    set(_include_paths "/usr/include" "/usr/src")
    set(_source_paths "/usr/src")
```

Removing the `/usr/src` allowed me to git rid of `/usr/include/{gtest, gmock}.h` and build successfully. I assume this is a mistake, either to have `/usr/src` there at all, or not using whatever cmake mechanism is necessary to set this as a list.

I now get this as the only relevant build output:
```
-- Forcing gtest/gmock from source, though one was otherwise available.
-- Found gtest sources under '/usr/src/gmock': gtests will be built
-- Found gmock sources under '/usr/src/gmock': gmock will be built
```

Now that I had a culprit, it explains the behavior I commented on above as well. By default, these are the [paths passed](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L181) to `_catkin_find_google_source()`:
```
### /usr/include;/usr/src/gtest;/usr/include;/usr/src/googletest/googletest/include/gtest
### /usr/src/gtest/src;/usr/src/googletest/googletest/src
### /usr/include;/usr/src/gmock;/usr/include;/usr/src/googletest/googlemock/include/gmock
### /usr/src/gmock/src;/usr/src/googletest/googlemock/src
```

As observed in a previous comment, `/usr/include/` is in there twice, which doesn't make any sense. Removing the trailing `/usr/src` entry [here](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L303) results in:
```
### /usr/include/gtest;/usr/include/googletest/googletest/include/gtest
### /usr/src/gtest/src;/usr/src/googletest/googletest/src
### /usr/include/gmock;/usr/include/googletest/googlemock/include/gmock
### /usr/src/gmock/src;/usr/src/googletest/googlemock/src
```
Thoughts based on this?

For the `src` issues, I could see having arch to match debian, but could also see adding the path here. Is there a preference?
```
#Find GTest
  set(_gtest_include_paths "${include_path}/gtest")
  set(_gtest_source_paths "${src_path}/gtest/src")
  list(APPEND _gtest_include_paths "${include_path}/googletest/googletest/include/gtest")
  list(APPEND _gtest_source_paths "${src_path}/googletest/googletest/src")
  ### add this, and analog for gmock
  list(APPEND _gtest_source_paths "${src_path}/gtest")  
```

For the `include` issues, I don't know cmake well enough to understand why `/usr/src` was in the `include_path`. The change came with [#1022](https://github.com/ros/catkin/pull/1022/files#diff-84d1585b7758a3723f688bcff3f8602cL296) where it went from `/usr/include/` to [both paths](https://github.com/ros/catkin/pull/1022/files#diff-84d1585b7758a3723f688bcff3f8602cR300)

I'd propose either removing it or using `list(APPEND)` instead of `set(foo bar)`.

---

Aside: is there a list of distros you ROS requires testing on before merging a PR? You stressed testing a couple times above, but in this case #1022 looks to have complicated things. In the end, it was probably that arch doesn't have `/usr/src/gmock/src/*.cc` and I only replicated `/usr/src/gmock/*`, not `/usr/src/gmock/src/*` in my previous attempt. Trying since #1022 brought about the path issue, so now there were two issues to debug.

I plan to hunt down the full story of what happened since 0.7.18 and now, then can post back with PR proposals.

**Comment by dirk-thomas** [comment]
> The change came with #1022 where it went from `/usr/include/` to both paths

Maybe @sloretz can comment on that who authored that change.

**Comment by dirk-thomas** [comment]
> is there a list of distros you ROS requires testing on before merging a PR?

Since `catkin` is the most low level package in ROS 1 the list of platforms a change can break is basically any platform someone might be building ROS on. That is why any change needs to be really clear *why* it is necessary and we need to be convinced that it doesn't break any potential use case.

**Comment by jwhendy** [comment]
> Maybe @sloretz can comment on that who authored that change.

It looks like he did, and then you merged it. I don't see any discussion of testing at all, nor does it look to refer to any issues so I'm not sure what specific issues prompted this.

I'm digging through internals more and am a bit out of my league to know what, exactly, needs fixing. I undid all my system changes, so I am back to the default file locations for [gmock](https://www.archlinux.org/packages/community/x86_64/gmock/) and [gtest](https://www.archlinux.org/packages/community/x86_64/gtest/) on arch linux.

My intuition was to change `_include_paths` to a list and append `/usr/src` vs. setting them at the same time. So, I started by changing [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303) from this:
```
    set(_include_paths "/usr/include" "/usr/src")
```
... to this:
```
    set(_include_paths "/usr/include")
    list(APPEND _include_paths "/usr/src")
```

I'm not sure what the desired outcome was, but `/usr/include/{gmock,gtest}` seems like it should be included, but the [package-specific dir here](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L258) is only appended to the last path.

Doing some output testing by modifying the [code here](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303):
```
    set(_include_paths "/usr/include" "/usr/src")
    set(_gmock_include_paths "${_include_paths}/gmock")
    message("### _gmock_include_paths: ${_gmock_include_paths}")
    ### _gmock_include_paths: /usr/include;/usr/src/gmock
```

Now, when I fix that to point at `/usr/include/gmock`, I get another error:
```
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:344 (add_subdirectory):
  The source directory

    /usr/src

  does not contain a CMakeLists.txt file.
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:346 (set_target_properties):
  set_target_properties Can not find target to add properties to: gtest
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:349 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


make: *** [Makefile:390: cmake_check_build_system] Error 1
```

Starting with the first, it's [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L341). Printing out the directories:
```
    message("### ${base_dir} ${gtest_lib_dir}")
    ### /usr/src /home/jwhendy/catkin_ws/build/catkin/gtest
```

I would not expect `CMakeLists.txt` in `/usr/src` either. If I manually change it to `/usr/src/gtest`, it goes away. Then I get an error at [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L346).
```
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:349 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
```
Adding some messages yields:
```
    message("### ${gtest_libs} ${gtest_main_libs}")
    message("### ${gmock_libs} ${gmock_main_libs}")
    ### gtest gtest_main
    ### gmock gmock_main
```

So, we're back to the filesystem woes:
```
# all good for gtest, hence no error
$ ls /usr/src/gtest/src/
gtest-all.cc  gtest.cc  gtest-death-test.cc  gtest-filepath.cc  gtest-internal-inl.h  gtest_main.cc  gtest-port.cc  gtest-printers.cc  gtest-test-part.cc  gtest-typed-test.cc

# booo
$ ls /usr/src/gmock/src
ls: cannot access '/usr/src/gmock/src': No such file or directory
```

So, I think one could make the case that arch should align these two packages to have the same `src` location.

I still think something is off, though. If I change [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L259) to this:
```
set(_gmock_source_paths "${src_path}/gmock")
```
That error still exists, even though that's where the files are. I then copied `*.cc` to `/usr/src/gmock/src` but am still getting the same error:
```
### gmock gmock_main
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:348 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
```

That's where I'll leave it for now. I feel extremely inefficient in tracking this down and there appears to be at least a couple independent sets of paths maintained depending on what goes hunting for `gtest`/`gmock`. This might make sense to someone, but I'm pretty confused on what ultimately needs to be found, which paths match up to what and for what purpose, and what's now breaking. This isn't the behavior I got yesterday and now I don't seem to be able to replicate my eventual success! Out of time for the day...

**Comment by dirk-thomas** [comment]
>  started by changing [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303) from this:
> ```
>     set(_include_paths "/usr/include" "/usr/src")
> ```
> ... to this:
> ```
>     set(_include_paths "/usr/include")
>     list(APPEND _include_paths "/usr/src")
> ```

Both code snippets are producing the very same result in CMake. Your code change is a no op. The variable was before and is afterwards a list with two paths.

**Comment by sloretz** [comment]
@jwhendy I'm unable to reproduce using this Dockerfile, but I'm not familiar with arch so I probably I made a mistake. Mind showing me what the correct steps are?

```Dockerfile
FROM archlinux/base

RUN pacman -Sy --noconfirm base-devel

# Add a user to build stuff
RUN useradd ros --create-home

# give user passwordless sudo access
RUN echo "ros ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

# Get yay
RUN pacman -Sy --noconfirm git go \
 && cd /tmp \
 && sudo -u ros git clone https://aur.archlinux.org/yay.git \
 && cd yay \
 && sudo -u ros makepkg -si --noconfirm

# Install ros-melodic-catkin with modifications to use 0.7.19
RUN sudo -u ros yay --useask -S --noconfirm gmock gtest python python-catkin_pkg python-empy python-nose ros-build-tools-py3 cmake \
 && cd /tmp \
 && sudo -u ros git clone https://github.com/ros-melodic-arch/ros-melodic-catkin.git \
 && cd ros-melodic-catkin \
 && sed -i.bak 's/0.7.18/0.7.19/g' .SRCINFO \
 && sed -i.bak 's/0.7.18/0.7.19/g' PKGBUILD \
 && sed -i.bak 's/b6a7428944911a011b0c3b0e465f2db04d219ce5f72cf3e2f76194dd055f1f49/b83d66640df99f72bc37160e8b60a76df6c87ff8dcbb9ab096911c44f08d13e1/' .SRCINFO \
 && sed -i.bak 's/b6a7428944911a011b0c3b0e465f2db04d219ce5f72cf3e2f76194dd055f1f49/b83d66640df99f72bc37160e8b60a76df6c87ff8dcbb9ab096911c44f08d13e1/' PKGBUILD \
 && sudo -u ros makepkg -si --noconfirm

# Try to build realtime tools
RUN sudo -u ros yay --useask -S --noconfirm ros-build-tools-py3 ros-melodic-realtime-tools
```


I guess this being at the end of the output means it succeeded?

```
Packages (1) ros-melodic-realtime-tools-1.13.1-0

Total Installed Size:  0.19 MiB

:: Proceed with installation? [Y/n] 
checking keyring...
checking package integrity...
loading package files...
checking for file conflicts...
checking available disk space...
:: Processing package changes...
installing ros-melodic-realtime-tools...
:: Running post-transaction hooks...
(1/1) Cleaning up package cache...

```

The package seems to exist
```
[root@62545e10fc86 /]# . /opt/ros/melodic/setup.bash
[root@62545e10fc86 /]# rospack find realtime_tools
/opt/ros/melodic/share/realtime_tools
```

**Comment by jwhendy** [comment]
@dirk-thomas There's kind of a lot of other stuff I mentioned in the past posts. I hope it's clear that I'm at the cmake noob level, so I'm not surprised that was incorrect.

I think it would be more helpful to focus on the heart of the issue vs. this goof up. As in, what did the rest of my post say about _why_ I made that change? It's because I assume `/usr/include/gmock` is path we want, but `_include_paths` is `/usr/include/;/usr/src` and then `gmock` is appended and we get `/usr/include/;/usr/src/gmock`. Would you kindly respond to that?

I hope I've shown my willingness to work on this, either by tracking it down here, getting arch to adjust, and/or PRs. So far, I have to say it's demotivating to put in a lot of effort trying stuff and to have responses focus on <1% of what I've actually said.

I offer free help, but I don't promise it will be good. If you value what I've contributed above, I'd love some comments after thorough processing of what I've found vs. just the mistakes. If you would like something different, please just let me know. I'm just looking for clues and posting what I find in hopes that the experts here can interpret it.

If what I'm doing is not of value, I can spend my time on other things.

**Comment by jwhendy** [comment]
@sloretz Nice! Thanks for trying this. The problem (or lack thereof), is that the version of `ros-melodic-realtime-tools` on AUR is still 1.13. `gmock` wasn't added until [1.14](https://github.com/ros-controls/realtime_tools/commit/2b9719650c44ef94b66580980f8b70a6dd1cbdfb).

I have never used docker, but I pushed a `src-fix` branch to [ros-melodic-realtime-tools](https://github.com/ros-melodic-arch/ros-melodic-realtime-tools/tree/src-fix) which would build 1.15.0. Can you try that instead?

Unfortunately, I'm a little hesitant to suggest how to replicate, as we are in an in between state: AUR still mostly points at `ros-gbp` releases, but we have updated nearly all packages in `ros-melodic-arch` [on github](https://github.com/ros-melodic-arch) to use the upstream sources directly.

When you use `yay`, you'll get AUR versions which *could* create goofs as a lot of package will be on older versions from `ros-gbp`.

---

Since you authored #1022, could you comment on what paths you believe should be searched in the functions I outlined in the past few comments? Specifically, is this what you expect/want include paths to be in `_catkin_find_google_source()` [here](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L181)?
```
/usr/include;/usr/src/gmock;/usr/include;/usr/src/googletest/googlemock/include/gmock
```

arch's `gmock` lives in `/usr/include/gmock`, which is also what @dirk-thomas found that debian's path [should be](https://packages.ubuntu.com/bionic/amd64/googletest/filelist), so I don't understand how it's found in the `include_paths` that looks to be used.

**Comment by dirk-thomas** [comment]
> That's where I'll leave it for now. I feel extremely inefficient in tracking this down and there appears to be at least a couple independent sets of paths maintained depending on what goes hunting for gtest/gmock. This might make sense to someone, but I'm pretty confused on what ultimately needs to be found, which paths match up to what and for what purpose, and what's now breaking. This isn't the behavior I got yesterday and now I don't seem to be able to replicate my eventual success!

> I hope I've shown my willingness to work on this, either by tracking it down here, getting arch to adjust, and/or PRs. So far, I have to say it's demotivating to put in a lot of effort trying stuff and to have responses focus on <1% of what I've actually said.

This sounds like you are implying that you have already done some work and are expecting that other will now jump in and help out on the missing parts. That is unfortunately not an expectation I can meet. My resources for ROS 1 maintenance are extremely limited and Arch is neither a supported platform nor have I ever used it myself. Maybe @sloretz will be able to continue helping on Arch specific debugging as he has done already. 

I am happy to comment were I can and where the information I am being given is sufficient for me to come up with a helpful opinion. If the information is not clear to me (since I might not understand what you are trying to describe or things are unclear to me / missing context) I am unfortunately unable to spend the necessary effort to fill these hole. Therefore some of my feedback will only be general debugging advice.

> It's because I assume `/usr/include/gmock` is path we want, but `_include_paths` is `/usr/include/;/usr/src` and then `gmock` is appended and we get `/usr/include/;/usr/src/gmock`. Would you kindly respond to that?

I agree with you that the current usage (passing a list of include paths but only appending a subdirectory to the last item) looks wrong. From your previous comment that conclusion was not that obvious to see - this statement was much more clear. See https://github.com/ros/catkin/pull/1022/files#r333317721 for a comment about it on the PR adding this.

For the next error message you are getting (`The source directory /usr/src does not contain a CMakeLists.txt file`) I simply don't have enough context / information to give you any useful feedback. There is certainly neither a `/usr/src/CMakeLists.txt` file on Ubuntu nor is it being attempted to find it there. Why is that happening on Arch? I don't know the answer to that - it might be related to previous modifications or some other difference between the two platforms.

Without trying it myself on Arch (which I don't have the bandwidth for) I can only give generic troubleshooting advice: I would advise to compare what happens on an Ubuntu machine (literally checking the complete flow of the CMake logic and the value of every variable) and compare that to what happens on Arch. That trace will not only uncover where those two platforms diverge but also improve the understanding what exactly happens and why.

> getting arch to adjust, and/or PRs.

I would suggest to you to separate these different aspects to simplify the problem to troubleshoot. I would first use e.g. symlinks to make the file system layout the same as in Debian and then try to figure out why it does work on Ubuntu but not on Arch (as you stated before). A complete trace as described in the previous paragraph will certainly provide that kind of information - it will take effort to collect though. Only once that has come to a conclusion it makes sense to consider if the logic should be updated to support the existing arch package or request updating the arch package to make it look like on Ubuntu.

> If what I'm doing is not of value, I can spend my time on other things.

We appreciate and value any kind of contribution. It would be great if you continue that work to resolve this problem on Arch.

**Comment by jwhendy** [comment]
>This sounds like you are implying that you have already done some work and are expecting that other will now jump in and help out on the missing parts. hat is unfortunately not an expectation I can meet.

Hopefully you agree I _have_ already done some work (as I've measured many hours of it), and I appreciate the correction on what you're able to provide. When I see an issue follow-up, I do tend to assume that a maintainer would like to resolve it.

>I am happy to comment were I can and where the information I am being given is sufficient for me to come up with a helpful opinion.

I think this is all I want as well. I'm happy to insert `message()` output to tell you what's going on, but learning `catkin` internals and `cmake` from scratch is also not an expectation I can meet in order to debug something. I'm contributing time out of charity to help experts, ideally, see something that stands out as weird. Then I have something to go by. Currently I've received no feedback on actual findings to say whether or not they look correct or not.

>From your previous comment that conclusion was not that obvious to see - this statement was much more clear. 

Yay! I like progress. Thanks for summarizing this nicely.

> I would advise to compare what happens on an Ubuntu machine (literally checking the complete flow of the CMake logic and the value of every variable) and compare that to what happens on Arch. 

I'm game, and dual boot bionic. At face value to a noob, I imagine putting `message()` lines everywhere... is there a more "pro" way to do said trace?

>I would suggest to you to separate these different aspects to simplify the problem to troubleshoot.

I mostly agree, especially after seeing `/usr/src/gtest/src` and *not* seeing this directory for `gmock`. That seems like a goof. That said, I was inclined to identify the minimum changes necessary to fix arch, then diagnose the best solution. For example, if other distros use `/usr/srcr/gmock/*.cc` (vs. putting in `src/*.cc`), maybe there is a case for just adding the path to existing paths (otherwise you'll duplicate this next time a non-debian distro pops up). In any case, the current path seems good and I'll summarize what I find.

> We appreciate and value any kind of contribution. It would be great if you continue that work to resolve this problem on Arch.

Your response this time seemed much more concrete and helpful, and I appreciate that. I didn't want to get into a cycle of posting a bunch of work that no one had any time to read, which is what previous interactions felt like. I really appreciate the pointers and am happy to continue tracking this down.

**Comment by gleichdick** [comment]
Hey folks, I have similar issues on openSuSE tumbleweed (the rolling release disro).

Basically I installed the gtest and gmock packages, the file layout is a bit different:
The headers are installed in `/usr/include/gtest` and `/usr/include/gmock`, the __source__ files are in `/usr/include/gtest/src` and `/usr/include/gmock/src`. Unfortunately, the `CMakeLists.txt` file is not provided.
But at least in tumbleweed, there is a CMake package in `/usr/lib64/cmake/GTest` which provides four targets (gtest, gmock and {gtest,gmock}_main) for modern style cmake.

So, catkin 0.7-19-1 only finds gtest, not gmock. 
As I didn't want to change system-wide directories, I edited `cmake/test/gtest.cmake` to point to a custom installation dir of googletest:
```cmake
# ll. 234, similar ll. 257 for gmock
  #Find GTest
  set(_gtest_include_paths "${include_path}/gtest")
  set(_gtest_source_paths "${src_path}/gtest/src")
  list(APPEND _gtest_include_paths "/home/user/ros/googletest/googletest/include/gtest")
  list(APPEND _gtest_source_paths "/home/user/ros/googletest/googletest/src")
  if(CATKIN_TOPLEVEL)
    # Ensure current workspace is searched before system path
    list(INSERT _gtest_include_paths 0 "/home/user/ros/googletest/googletest/include/gtest")
    list(INSERT _gtest_source_paths 0 "/home/user/ros/googletest/googletest/src")
  endif()
```

So I'd suggest to introduce a new CMake variable for overriding the gtest source dir. Anyway, I'm very happy to help :smile:

**Comment by jwhendy** [comment]
I spent some time troubleshooting again. I am still a bit perplexed, as I don't think my experience replicates previous attempts, but oh well. I needed the following to build `realtime_tools`:

- link `*.cc` from `/usr/src/gmock` (as installed) to `/usr/src/gmock/src`; I will followup on [my existing bug report](https://bugs.archlinux.org/task/63793) to arch linux to get this to happen
- remove the `/usr/src` from [L303](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303). I will defer to someone else on how to get `gmock` appended to all paths in `include_dirs` vs. the current behavior.
- I needed to [download my own](https://github.com/google/googletest/blob/master/googlemock/CMakeLists.txt) `/usr/src/gmock/CMakeLists.txt`, as it appears it's not provided by arch linux, though there is a `/usr/src/gtest/CMakeLists.txt`.
- as is, the `CMakeLists.txt` above doesn't work. I had to change [L17](https://github.com/google/googletest/blob/master/googlemock/CMakeLists.txt#L17) and guessed at using `../gtest` instead of `../googletest`, but have no idea...

This might be related to the last bullet, but I now get this when building `realtime_tools`:

```
CMake Warning at /usr/src/gmock/CMakeLists.txt:37 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


CMake Warning (dev) at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  Policy CMP0013 is not set: Duplicate binary directories are not allowed.
  Run "cmake --help-policy CMP0013" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The binary directory

    /home/jwhendy/catkin_ws/build/realtime_tools/gtest

  is already used to build a source directory.  This command uses it to build
  source directory

    /usr/src/gtest

  which can generate conflicting build files.  CMake does not support this
  use case but it used to work accidentally and is being allowed for
  compatibility.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning at /usr/src/gtest/CMakeLists.txt:47 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.
```

For my part, I think the following are the actions I propose:
- [ ] the team here sorts out how [L303](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303) is supposed to be handled
- [ ] I'll pursue getting `/usr/src/gmock/` files sorted out with arch linux

**Comment by mikepurvis** [comment]
We're seeing this at Clearpath as well when we switch to catkin 0.7.19— our platform is Ubuntu 16.04. Our gtest is just the one from upstream, `libgtest-dev` at version `1.7.0-4ubuntu1`, file list:

https://packages.ubuntu.com/xenial/amd64/libgtest-dev/filelist

----

I can also confirm that changing

https://github.com/ros/catkin/blob/d564828ac55163ca42611e9ac0fd9d11a1d1cca4/cmake/test/gtest.cmake#L303-L304

To just be this fixes it for us:

```
    set(_include_paths "/usr/include")
    set(_source_paths "/usr/src")
```

**Comment by mikepurvis** [comment]
Following from the discussion in https://github.com/ros/catkin/pull/1022#discussion_r333317721, unfortunately, `list(TRANSFORM APPEND)` isn't in CMake 3.5.1, so append-to-each-item must be done manually without it as long as we're supporting 16.04 from this branch. But I think it's more complicated than that because those lists are appended-to _twice_, see:

https://github.com/ros/catkin/blob/d564828ac55163ca42611e9ac0fd9d11a1d1cca4/cmake/test/gtest.cmake#L235-L238

Without really fully understanding what motivated the original change from @sloretz in #1022, my instinct is to leave `include_path` and `src_path` as non-list parameters to the function, and instead do the expansion as part of these other operations, eg something like:

```
set(_gtest_include_paths "${include_path}/gtest")
set(_gtest_source_paths "${src_path}/gtest/src")
list(APPEND _gtest_include_paths "${src_path}/googletest/googletest/include/gtest")
list(APPEND _gtest_source_paths "${src_path}/googletest/googletest/src")
```

Or just add a third parameter to the function for the location of googletest, so it's more like:

```
set(_gtest_include_paths "${include_path}/gtest")
set(_gtest_source_paths "${src_path}/gtest/src")
list(APPEND _gtest_include_paths "${googletest_path}/googletest/include/gtest")
list(APPEND _gtest_source_paths "${googletest_path}/googletest/src")
```

IMO #1022 should be reverted until someone with a Buster setup can find the right combination of the above that makes it work in that scenario without breaking everyone else.

**Comment by jwhendy** [comment]
@dirk-thomas follow-up on the src warnings. As mentioned above, this line in [the default gmock `CMakeLists.txt`](https://github.com/google/googletest/blob/master/googlemock/CMakeLists.txt) does not work:

```
# A directory to find Google Test sources.
if (EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/gtest/CMakeLists.txt")
  set(gtest_dir gtest)
else()
  set(gtest_dir ../googletest)
endif()
```

With this default, I get:
```
Starting  >>> catkin                                                                                                                                                                     
_________________________________________________________________________________________________________________________________________________________________________________________
Errors     << catkin:check /home/jwhendy/catkin_ws/logs/catkin/build.check.002.log                                                                                                       
CMake Warning at /usr/src/gmock/CMakeLists.txt:37 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


CMake Error at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  add_subdirectory given source "../googletest" which is not an existing
  directory.


CMake Error at /usr/src/gmock/CMakeLists.txt:59 (config_compiler_and_linker):
  Unknown CMake command "config_compiler_and_linker".


make: *** [Makefile:446: cmake_check_build_system] Error 1
```

I didn't know what to make this, so I guessed `../gtest` instead. That does build, but leads to a warning for every package like this:
```
Starting  >>> catkin                                                                                                                                                                     
_________________________________________________________________________________________________________________________________________________________________________________________
Warnings   << catkin:check /home/jwhendy/catkin_ws/logs/catkin/build.check.003.log                                                                                                       
CMake Warning at /usr/src/gmock/CMakeLists.txt:37 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


CMake Warning (dev) at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  Policy CMP0013 is not set: Duplicate binary directories are not allowed.
  Run "cmake --help-policy CMP0013" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The binary directory

    /home/jwhendy/catkin_ws/build/catkin/gtest

  is already used to build a source directory.  This command uses it to build
  source directory

    /usr/src/gtest

  which can generate conflicting build files.  CMake does not support this
  use case but it used to work accidentally and is being allowed for
  compatibility.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning at /usr/src/gtest/CMakeLists.txt:47 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.
```

I'm guessing the proposal would be to take this up with googletest and arch?

**Edit:** for clarification, arch doesn't ship with a `/usr/src/gmock/CMakeLists.txt` for whatever reason. So, it will break using the default package, then also break of you google around for what to do about this missing file (i.e. download it), and give warnings when I do the above.

**Comment by sloretz** [comment]
I don't understand all of the gtest/gmock logic, but I had a look through it. Hopefully this is useful. IIUC catkin first tries to decide whether to `find_package()` GTest/GMock, or build it from source. If it decides the latter, then it tries to build from source in the same project.

#1022 is supposed to just refactor the logic that makes the decision (catkin was choosing to use GTest from the find module when it should have built gtest/gmock from source), but as reported by @mikepurvis it seems to have broken the code that builds gtest/gmock on Xenial. #1040 doesn't touch the decision logic, but instead fixes the finding of gtest/gmock sources on Jessie/Xenial.

Neither #1022 or #1040 address Arch (this ticket). I don't know if it's the decision logic or the code that finds gtest/gmock sources that needs to be fixed for Arch, but I suspect it's the latter based on @jwhendy's note:

> Edit: for clarification, arch doesn't ship with a /usr/src/gmock/CMakeLists.txt for whatever reason.

If it got that far then there weren't find modules for both GTest and GMock, so it tried to build from source. The logic to build gtest/gmock assumes there is a `CMakeLists.txt` and [calls `add_subdirectory()`](https://github.com/ros/catkin/blob/f7429885cf5bef32265626f6eefac78ae263daf1/cmake/test/gtest.cmake#L341) to make the build of gtest/gmock happen. I'm not sure about using `add_subdirectory()` as the tool for building `gtest/gmock`, but it seems to be [recommended by googletest](https://github.com/google/googletest/blob/master/googletest/README.md#incorporating-into-an-existing-cmake-project). For that to succeed, there must a directory containing a `CMakeLists.txt` for gtest and and another directory containing a `CMakeLists.txt` for gmock.

IIUC `catkin_find_google_test_source()` is supposed to find a directory containing a `CMakeLists.txt` for gtest or gmock. It does this by looking for source code, and then [assumes a directory above that will have a `CMakeLists.txt`](https://github.com/ros/catkin/blob/f7429885cf5bef32265626f6eefac78ae263daf1/cmake/test/gtest.cmake#L195-L196), which is returned as [`base_dir` and given to `add_subdirectory()`](https://github.com/ros/catkin/blob/f7429885cf5bef32265626f6eefac78ae263daf1/cmake/test/gtest.cmake#L341). I don't know why it doesn't look for `CMakeLists.txt` directly. I don't know why it looks for the header files and sets `GTEST/GMOCK_INCLUDE_DIRS` directly. I would think header file locations could be retrieved from targets or variables after the call to `add_subdirectory()`.

IIUC the locations `catkin_find_google_test_source()` should come up with for `base_dir` are these directories:

gtest CMakeList.txt:
* [jessie](https://packages.debian.org/jessie/amd64/libgtest-dev/filelist), [xenial](https://packages.ubuntu.com/xenial/amd64/libgtest-dev/filelist)
  * `/usr/src/gtest/CMakeLists.txt`
* [stretch](https://packages.debian.org/stretch/amd64/googletest/filelist), [bionic](https://packages.ubuntu.com/bionic/amd64/googletest/filelist), [buster](https://packages.debian.org/buster/amd64/googletest/filelist):
  * `/usr/src/googletest/googletest/CMakeLists.txt`
* arch
  * ???

gmock CMakeList.txt:
* [jessie](https://packages.debian.org/jessie/amd64/google-mock/filelist), [xenial](https://packages.ubuntu.com/xenial/amd64/google-mock/filelist)
  * `/usr/src/gmock/CMakeLists.txt`
* [stretch](https://packages.debian.org/stretch/amd64/googletest/filelist), [bionic](https://packages.ubuntu.com/bionic/amd64/googletest/filelist), [buster](https://packages.debian.org/buster/amd64/googletest/filelist):
  * `/usr/src/googletest/googlemock/CMakeLists.txt`
* arch
  * ???

**Comment by sloretz** [comment]
Actually, it looks even more complicated. It looks like `catkin_find_google_test_source()` returns only a single base directory for both gtest and gmock depending on what sources are found. For stretch/bionic/buster it looks like this is `/usr/src/googletest/CMakeLists.txt`. For jessie/xenial, I'm not sure. It looks like it uses the gmock CMakeLists.txt? Is there an assumption that gmock provides gtest too?

**Comment by sloretz** [comment]
Reopening since I don't think #1040 is expected to resolve the issue on Arch.

**Comment by jwhendy** [comment]
Thanks for the analysis @sloretz ! I think at this point, the `catkin` related things are solved and this is on me to take up with arch. That said, I still feel out of my league so any pointers on what might be awry on arch would still be highly appreciated (even though this could probably be closed as unrelated now).

Thus far:
- [ ] for sure we need `/usr/src/gmock/src/*.cc` to exist (hope to accomplish via [this bug request](https://bugs.archlinux.org/task/63793))
- [ ] ideally, include a functioning `CMakeLists.txt` with the `gmock` package

For arch's files,
- [gtest](https://www.archlinux.org/packages/community/x86_64/gtest/): `usr/src/gtest/CMakeLists.txt`
- none shipped with gmock

I tried [this file](https://www.apt-browse.org/browse/ubuntu/trusty/main/amd64/google-mock/1.6.0+svn437-0ubuntu5/file/usr/src/gmock/CMakeLists.txt) from Ubuntu Trusty and no longer get the errors. It points to a `../gtest` directory, but `bionic` points to `../googletest`. Diffing vs. the default file from `google` has quite a bit of differences. I'm not sure how custom these files are to the distribution or what.

**Comment by liborw** [comment]
Hi, [here](https://gist.github.com/liborw/2af80305e3005184a21a576ba9913891) is a PKGBUILD for the googletest that I have used to install both gtest and gmock version 1.8.1, which hopefully provides all the necessary files in order to be found by catkin. The realtime-tools then compile without problems.

There is just small problem that there is no VERSION in the CMakeLists and catkin is complaining.

**Comment by mikepurvis** [comment]
@liborw Can you try again as of https://github.com/ros/catkin/pull/1052, and see if the `VERSION`-related warning is now silenced?

**Comment by acxz** [comment]
Should be resolved with the merge of this PR: anthraxx/arch-pkgbuilds#33. Since no change needs to be made in this codebase, feel free to close this issue.

**Comment by dirk-thomas** [comment]
> Since no change needs to be made in this codebase, feel free to close this issue.

Sounds good to me. Please feel free to comment on the closed ticket and it can be reopened if necessary.

**Comment by acxz** [comment]
Did some more thorough testing and even though catkin now builds on Arch with the above linked PR, realtime-tools still does not. Sorry for the haste, request to reopen this issue.

**Comment by jwhendy** [comment]
@acxz any notable error messages you could post?

**Comment by acxz** [comment]
The error message is the exact same as in the OP. When I install catkin it identifies gmock and gtest but when I want to use those catkin cmake functions it chokes. Sadly I don’t have time to spend on this further. One thing I want to try once I do get some time or if someone wants to PR, is change how gtest/gmock is installed on Arch to *exactly* match the ubuntu package hierarchy. If that doesn’t fix it, I’ll admit defeat.

**Comment by jwhendy** [comment]
Ok, just trying to understand and I can look into this later this week/weekend.

For reference, can you?
```
ls /usr/src/gmock/src
```

**Comment by acxz** [comment]
```
acxz@archard ~ (git)-[master] % ls /usr/src/gmock/src
gmock.cc      gmock-cardinalities.cc   gmock-matchers.cc       gmock_main.cc
gmock-all.cc  gmock-internal-utils.cc  gmock-spec-builders.cc
```

**Comment by jjaaseon** [comment]
I recently encountered the same problem on Manjaro.

I manually compiled ros-melodic-realtime-tools and found that the problem was [GMOCK_MAIN_LIBRARIES ][1]. CATKIN_ENABLE_TESTING is "ON" while GMOCK_MAIN_LIBRARIES was empty. Then I checked the [CMake Module][2] and found that FindGTest is already shipped with CMake but not FindGMock. 

Here is my solution:
- Take the FindGMock.cmake script from the net (e.g.,[FindGMock.cmake][3]) and copy that script into my project.
- Then add the following in the PKGBUILD of the ros-melodic-realtime-tools
		-DCMAKE_MODULE_PATH=/FindGMock.cmake/Modules/Path
- Then use makepkg to compile manually, and finally succeeded！

Hope it would help.

[1]: https://github.com/ros-controls/realtime_tools/blob/33893cd36ded92d167ed9b6e0d9074f1b543fffc/CMakeLists.txt#L24
[2]: https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html
[3]: https://gitlab.enseeiht.fr/abcd/abcd/blob/mumps_scaling/cmake/Modules/FindGMock.cmake

**Comment by acxz** [comment]
@liborw you mentioned that your PKGBUILD solves building realtime-tools, but I tested it out and it did not work for me.

**Comment by acxz** [comment]
I recreated the Ubuntu installation of gtest and gmock as shown [here](https://packages.ubuntu.com/bionic/amd64/googletest/filelist). You can see my attempts [here](https://aur.archlinux.org/pkgbase/gtest-py3/) at the AUR `gtest-py3/gmock-py3` package.
However, I was still unsuccessful.

**Comment by acxz** [comment]
@jiaaseon mentioned something important that the GMOCK_MAIN_LIBRARIES variable is empty. Based on that advice I am patching the AUR package for realtime-tools to not test if that variable is empty. https://github.com/ros-melodic-arch/ros-melodic-realtime-tools/pull/5

Also submitted a PR to realtime-tools with the patch: https://github.com/ros-controls/realtime_tools/pull/55

**Comment by acxz** [comment]
Finally got around to resolving this. Submitted a patch to the Arch Linux repos to get gtest updated to 1.10.0 and restructure the installed files based on upstream recommendations. Just need to add one more path for the googletest sources see above linked PR in catkin's gtest.cmake and we good to go.

**Comment by LinuxPersonEC** [comment]
I'm seeing this on Fedora 32 with ROS Noetic as well. Which fix works?

```
rpm -qa|grep gtest
gtest-1.10.0-1.fc32.x86_64
gtest-devel-1.10.0-1.fc32.x86_64
```


```
-- ==> add_subdirectory(realtime_tools)
CMake Warning at /root/ros_catkin_ws/install_isolated/share/catkin/cmake/test/gtest.cmake:160 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /root/ros_catkin_ws/install_isolated/share/catkin/cmake/test/gtest.cmake:89 (_catkin_add_executable_with_google_test)
  /root/ros_catkin_ws/install_isolated/share/catkin/cmake/test/gtest.cmake:63 (_catkin_add_google_test)
  realtime_tools/CMakeLists.txt:23 (catkin_add_gmock)

CMake Error at realtime_tools/CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.

-- Configuring incomplete, errors occurred!
See also "/root/catkin_ws/build/CMakeFiles/CMakeOutput.log".
See also "/root/catkin_ws/build/CMakeFiles/CMakeError.log".
make: *** [Makefile:284: cmake_check_build_system] Error 1
Invoking "make cmake_check_build_system" failed
```
